File: ios/RecipeClipper/UI/Recipe/CookView.swift
import SwiftUI
import UIKit

private enum StepStatus { case done, current, upcoming }

/// A highlighted scroll, not a pager: cooking isn't linear, so the next step has to be
/// glanceable before it is current, and a scroll still works when a source gives three
/// paragraph-blobs instead of twelve clean steps.
struct CookView: View {
    let content: RecipeSuccess
    let state: RecipeUiState
    let vm: RecipeViewModel

    var body: some View {
        let cook = state.cook
        let steps = content.instructions

        VStack(spacing: 0) {
            CookTopBar(
                title: content.recipe.name,
                position: Strings.cookPosition(cook.currentStep + 1, of: steps.count),
                onExit: vm.onCookExit
            )
            IngredientsBar(content: content, state: state, vm: vm)
                // Sized ahead of the steps' scroll view, so an expanded list gets its full
                // allowance at the larger text sizes instead of an even share of the height.
                .layoutPriority(1)

            ScrollViewReader { proxy in
                ScrollView {
                    VStack(spacing: 10) {
                        ForEach(Array(steps.indices), id: \.self) { index in
                            CookStep(
                                index: index,
                                // Chef mode (#100): the short version, unless asked for as written.
                                text: content.shownStep(index, asWritten: state.asWrittenSteps),
                                amounts: content.shownStepParts(index, asWritten: state.asWrittenSteps),
                                shortToggle: !content.hasShortStep(index) ? nil
                                    : state.asWrittenSteps.contains(index) ? Strings.stepShowShort : Strings.stepShowAsWritten,
                                status: index == cook.currentStep ? .current
                                    : cook.doneSteps.contains(index) ? .done : .upcoming,
                                timerSeconds: index < content.stepTimerSeconds.count ? content.stepTimerSeconds[index] : nil,
                                // A step has a timer only when the app has the recipe's words.
                                timerWords: content.words ?? .english,
                                timer: cook.timers[index],
                                isLast: index == steps.count - 1,
                                vm: vm
                            )
                            .id(index)
                        }
                    }
                    .padding(.horizontal, 20)
                    .readableColumn()
                    .padding(.top, 12)
                    .padding(.bottom, 64)
                }
                .onAppear { proxy.scrollTo(cook.currentStep, anchor: .top) }
                .onChange(of: cook.currentStep) { _, step in
                    withAnimation { proxy.scrollTo(step, anchor: .top) }
                }
            }
        }
        // Cook mode holds the screen awake: nobody wants it dimming with flour on their hands.
        .keepsScreenOn()
    }
}

/// Android's FLAG_KEEP_SCREEN_ON, scoped to cook mode being on screen. A count of visible
/// cook views rather than a plain on/off: popping one recipe screen that was cooking back to
/// another that is cooking runs the lower one's onAppear *before* the upper one's
/// onDisappear, and a boolean set in each would end up off while cook mode is showing.
/// The idle timer is also released while the app is not active and taken back on return.
@MainActor
enum KeepScreenOn {
    private(set) static var holders = 0
    private(set) static var appActive = true

    static func acquire() { holders += 1; apply() }
    static func release() { holders = max(0, holders - 1); apply() }
    static func setAppActive(_ active: Bool) { appActive = active; apply() }

    /// What `isIdleTimerDisabled` should be.
    static var wanted: Bool { holders > 0 && appActive }

    private static func apply() {
        UIApplication.shared.isIdleTimerDisabled = wanted
    }

    /// For tests.
    static func reset() { holders = 0; appActive = true; apply() }
}

private struct KeepsScreenOn: ViewModifier {
    @Environment(\.scenePhase) private var scenePhase

    func body(content: Content) -> some View {
        content
            .onAppear {
                KeepScreenOn.setAppActive(scenePhase == .active)
                KeepScreenOn.acquire()
            }
            .onDisappear { KeepScreenOn.release() }
            .onChange(of: scenePhase) { _, phase in KeepScreenOn.setAppActive(phase == .active) }
    }
}

extension View {
    func keepsScreenOn() -> some View { modifier(KeepsScreenOn()) }
}

private struct CookTopBar: View {
    let title: String
    let position: String
    let onExit: () -> Void
    @Environment(\.dynamicTypeSize) private var dynamicTypeSize

    var body: some View {
        // At the accessibility sizes Exit and the counter stack: side by side, German's
        // "Beenden" and "Schritt 1 von 2" were each broken mid-word.
        let layout = dynamicTypeSize.isAccessibilitySize
            ? AnyLayout(VStackLayout(alignment: .leading, spacing: 0))
            : AnyLayout(HStackLayout(spacing: 8))
        layout {
            Button(Strings.exit, action: onExit)
                .buttonStyle(TextActionStyle(color: Palette.muted))
            // At the accessibility sizes the title goes: it is the least useful of the three
            // while cooking (it was on the screen you came from).
            if !dynamicTypeSize.isAccessibilitySize {
                Text(title)
                    .textStyle(Typography.titleSmall)
                    .foregroundStyle(Palette.onBackground)
                    .lineLimit(1)
                    .frame(maxWidth: .infinity, alignment: .leading)
            }
            Text(position)
                .textStyle(Typography.labelMedium)
                .foregroundStyle(Palette.muted)
                .padding(.leading, dynamicTypeSize.isAccessibilitySize ? 12 : 0)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.leading, 8)
        .padding(.trailing, 20)
        // Exit's own 12pt padding puts its text on the 20pt gutter, so this lines up with
        // the column below.
        .readableColumn()
        .padding(.vertical, 4)
    }
}

/// Ingredients stay one tap away, folded up so the steps own the screen.
private struct IngredientsBar: View {
    let content: RecipeSuccess
    let state: RecipeUiState
    let vm: RecipeViewModel
    @Environment(\.dynamicTypeSize) private var dynamicTypeSize
    /// Room for more ingredients when each one is bigger, capped so the steps keep the rest.
    @ScaledMetric(relativeTo: .body) private var scaledListHeight: CGFloat = 260
    private var listHeight: CGFloat { min(scaledListHeight, 420) }

    var body: some View {
        let expanded = state.cook.ingredientsExpanded
        VStack(spacing: 0) {
            Hairline()
            Button(action: vm.onIngredientsToggle) {
                HStack(spacing: 0) {
                    // The count drops under the heading at the accessibility sizes instead of
                    // wrapping mid-phrase beside it.
                    let stacked = dynamicTypeSize.isAccessibilitySize
                    let count = content.ingredients.count
                    let heading = stacked
                        ? AnyLayout(VStackLayout(alignment: .leading, spacing: 0))
                        : AnyLayout(HStackLayout(spacing: 0))
                    heading {
                        Text(Strings.headingIngredients)
                            .textStyle(Typography.titleSmall)
                            .foregroundStyle(Palette.onBackground)
                        Text(stacked ? Strings.cookIngredientsCountOwnLine(count) : Strings.cookIngredientsCount(count))
                            .textStyle(Typography.bodyMedium)
                            .foregroundStyle(Palette.muted)
                    }
                    Spacer(minLength: 8)
                    Text(expanded ? "▴" : "▾")
                        .textStyle(Typography.bodyLarge)
                        .foregroundStyle(Palette.muted)
                }
                .padding(.horizontal, 20)
                .readableColumn()
                .padding(.vertical, 14)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityHint(expanded ? Strings.hideIngredients : Strings.showIngredients)

            if expanded {
                let list = VStack(spacing: 0) {
                    ForEach(Array(content.ingredients.enumerated()), id: \.offset) { index, ingredient in
                        IngredientRow(
                            text: ingredient,
                            checked: state.checkedIngredients.contains(index),
                            onCheckedChange: { vm.onIngredientChecked(index, $0) }
                        )
                    }
                }
                .padding(.horizontal, 20)
                .readableColumn()
                .padding(.bottom, 8)

                ViewThatFits(in: .vertical) {
                    list
                    ScrollView { list }
                }
                .frame(maxHeight: listHeight)
            }
            Hairline()
        }
    }
}

private struct CookStep: View {
    let index: Int
    let text: String
    let amounts: [StepAmounts.Part]?
    /// Chef mode (#100): the current step's "As written" / "Short version" button; nil: none. A
    /// tap on a step already makes it current, so the switch is a small button on the card.
    let shortToggle: String?
    let status: StepStatus
    let timerSeconds: Int?
    let timerWords: LanguageWords
    let timer: StepTimer?
    let isLast: Bool
    let vm: RecipeViewModel
    @ScaledMetric(relativeTo: .headline) private var numberColumn: CGFloat = 32
    @Environment(\.dynamicTypeSize) private var dynamicTypeSize

    var body: some View {
        if status == .current {
            VStack(alignment: .leading, spacing: 0) {
                HStack(alignment: .firstTextBaseline) {
                    Text(Strings.cookStepLabel(index + 1))
                        .textStyle(Typography.labelSmall)
                        .foregroundStyle(Palette.accentText)
                    Spacer()
                    if let shortToggle {
                        Button(shortToggle) { vm.onStepAsWrittenToggle(index) }
                            .textStyle(Typography.labelMedium)
                            .foregroundStyle(Palette.accentText)
                    }
                }
                .padding(.bottom, 8)
                stepText(text, amounts, accent: Palette.accentText)
                    .textStyle(Typography.cookStep)
                    .foregroundStyle(Palette.onBackground)
                    .frame(maxWidth: .infinity, alignment: .leading)
                if timerSeconds != nil || timer != nil {
                    CurrentTimer(step: index, timerSeconds: timerSeconds, timerWords: timerWords, timer: timer, vm: vm)
                        .padding(.top, 16)
                }
                Button(isLast ? Strings.cookDoneFinish : Strings.cookDoneNext, action: vm.onStepDone)
                    .buttonStyle(PrimaryButtonStyle(minHeight: 52, fillWidth: true))
                    .padding(.top, 20)
            }
            .padding(20)
            .overlay(RoundedRectangle(cornerRadius: 18).strokeBorder(Palette.primary, lineWidth: 2))
        } else {
            let done = status == .done
            Button { vm.onStepSelected(index) } label: {
                // As in the reading view: the number goes above the step at the
                // accessibility sizes rather than taking a third of the width beside it.
                let stacked = dynamicTypeSize.isAccessibilitySize
                let layout = stacked
                    ? AnyLayout(VStackLayout(alignment: .leading, spacing: 0))
                    : AnyLayout(HStackLayout(alignment: .firstTextBaseline, spacing: 0))
                VStack(alignment: .leading, spacing: 4) {
                    layout {
                        Text("\(index + 1)")
                            .textStyle(Typography.titleMedium)
                            .foregroundStyle(done ? Palette.muted : Palette.accentText)
                            .frame(width: stacked ? nil : numberColumn, alignment: .leading)
                        // A done step is dimmed as a whole; its amounts keep only their weight.
                        stepText(text, amounts, accent: done ? nil : Palette.accentText)
                            .textStyle(Typography.bodyLarge)
                            .strikethrough(done)
                            .foregroundStyle(Palette.muted)
                            .multilineTextAlignment(.leading)
                            .frame(maxWidth: .infinity, alignment: .leading)
                    }
                    // A timer still running on a step you've moved on from stays visible:
                    // steps overlap.
                    if let timer {
                        Text(Strings.timerRunning(timer.finished ? Strings.timesUp : StepTimers.clock(timer.remainingSeconds)))
                            .textStyle(Typography.labelLarge)
                            .monospacedDigit()
                            .foregroundStyle(Palette.accentText)
                            .padding(.leading, stacked ? 0 : numberColumn)
                    }
                }
                .padding(.horizontal, 4)
                .padding(.vertical, 6)
                .opacity(done ? 0.45 : 1)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityHint(Strings.goToStep(index + 1))
        }
    }
}

private struct CurrentTimer: View {
    let step: Int
    let timerSeconds: Int?
    let timerWords: LanguageWords
    let timer: StepTimer?
    let vm: RecipeViewModel

    var body: some View {
        if let timer {
            // The clock and its buttons share a row while they fit; at the larger sizes the
            // buttons go under the clock (and under each other) rather than squeezing it.
            ViewThatFits(in: .horizontal) {
                HStack(spacing: 0) {
                    clock(timer).fixedSize()
                    Spacer()
                    buttons(timer).fixedSize()
                }
                VStack(alignment: .leading, spacing: 4) {
                    clock(timer)
                    HStack(spacing: 0) { buttons(timer).fixedSize() }.padding(.leading, -12)
                }
                VStack(alignment: .leading, spacing: 4) {
                    clock(timer)
                    VStack(alignment: .leading, spacing: 0) { buttons(timer) }.padding(.leading, -12)
                }
            }
        } else if let timerSeconds {
            Button(Strings.timerStart(StepTimers.label(timerSeconds, words: timerWords))) { vm.onTimerStart(step) }
                .buttonStyle(OutlinedActionStyle())
        }
    }

    private func clock(_ timer: StepTimer) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(StepTimers.clock(timer.remainingSeconds))
                .textStyle(Typography.timerClock)
                .foregroundStyle(timer.finished ? Palette.accentText : Palette.onBackground)
            if timer.finished {
                Text(Strings.timesUp)
                    .textStyle(Typography.labelMedium)
                    .foregroundStyle(Palette.accentText)
            }
        }
    }

    @ViewBuilder
    private func buttons(_ timer: StepTimer) -> some View {
        if !timer.finished {
            Button(timer.running ? Strings.pause : Strings.resume) { vm.onTimerToggle(step) }
                .buttonStyle(TextActionStyle())
        }
        Button(Strings.reset) { vm.onTimerReset(step) }
            .buttonStyle(TextActionStyle(color: Palette.muted))
    }
}

File: app/src/main/java/com/example/recipeclipper/ui/recipe/CookView.kt
package com.example.recipeclipper.ui.recipe

import android.view.WindowManager
import androidx.activity.compose.BackHandler
import androidx.activity.compose.LocalActivity
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.recipeclipper.R
import com.example.recipeclipper.data.model.LanguageWords
import com.example.recipeclipper.data.model.StepAmounts
import com.example.recipeclipper.data.model.StepTimers

private enum class StepStatus { DONE, CURRENT, UPCOMING }

/**
 * A highlighted scroll, not a pager: cooking isn't linear, so the next step has to be
 * glanceable before it is current, and a scroll still works when a source gives three
 * paragraph-blobs instead of twelve clean steps.
 */
@Composable
internal fun CookView(
    content: RecipeContent.Success,
    state: RecipeUiState,
    actions: RecipeActions
) {
    val cook = state.cook
    val steps = content.instructions
    val listState = rememberLazyListState()

    KeepScreenOn()
    BackHandler(onBack = actions.onCookExit)
    LaunchedEffect(cook.currentStep) {
        if (steps.isNotEmpty()) listState.animateScrollToItem(cook.currentStep.coerceIn(0, steps.lastIndex))
    }

    Column(Modifier.fillMaxSize()) {
        CookTopBar(
            title = content.recipe.name,
            position = stringResource(R.string.cook_position, cook.currentStep + 1, steps.size),
            onExit = actions.onCookExit
        )
        IngredientsBar(content, state, actions)

        LazyColumn(
            state = listState,
            contentPadding = PaddingValues(start = 20.dp, end = 20.dp, top = 12.dp, bottom = 64.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
            modifier = Modifier.weight(1f)
        ) {
            itemsIndexed(steps) { index, text ->
                val status = when {
                    index == cook.currentStep -> StepStatus.CURRENT
                    index in cook.doneSteps -> StepStatus.DONE
                    else -> StepStatus.UPCOMING
                }
                CookStep(
                    index = index,
                    text = text,
                    amounts = content.stepAmounts?.getOrNull(index),
                    status = status,
                    timerSeconds = content.stepTimerSeconds.getOrNull(index),
                    // A step has a timer only when the app has the recipe's words.
                    timerWords = content.words ?: LanguageWords.ENGLISH,
                    timer = cook.timers[index],
                    isLast = index == steps.lastIndex,
                    actions = actions
                )
            }
        }
    }
}

@Composable
private fun CookTopBar(title: String, position: String, onExit: () -> Unit) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier.fillMaxWidth().padding(start = 8.dp, end = 20.dp)
    ) {
        TextButton(onClick = onExit) {
            Text(
                stringResource(R.string.action_exit),
                style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
        Text(
            title,
            style = MaterialTheme.typography.titleSmall,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.weight(1f).padding(horizontal = 8.dp)
        )
        Text(
            position,
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}

/** Ingredients stay one tap away, folded up so the steps own the screen. */
@Composable
private fun IngredientsBar(
    content: RecipeContent.Success,
    state: RecipeUiState,
    actions: RecipeActions
) {
    val expanded = state.cook.ingredientsExpanded
    Column {
        Hairline()
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier
                .fillMaxWidth()
                .clickable(
                    role = Role.Button,
                    onClickLabel = stringResource(
                        if (expanded) R.string.cd_hide_ingredients else R.string.cd_show_ingredients
                    ),
                    onClick = actions.onIngredientsToggle
                )
                .padding(horizontal = 20.dp, vertical = 14.dp)
        ) {
            Text(stringResource(R.string.heading_ingredients), style = MaterialTheme.typography.titleSmall)
            Text(
                stringResource(R.string.cook_ingredients_count, content.ingredients.size),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
            Spacer(Modifier.weight(1f))
            Text(
                if (expanded) "▴" else "▾",
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
        if (expanded) {
            Column(
                Modifier
                    .heightIn(max = 260.dp)
                    .verticalScroll(rememberScrollState())
                    .padding(horizontal = 20.dp)
            ) {
                content.ingredients.forEachIndexed { index, ingredient ->
                    IngredientRow(
                        text = ingredient,
                        checked = index in state.checkedIngredients,
                        onCheckedChange = { actions.onIngredientChecked(index, it) }
                    )
                }
            }
        }
        Hairline()
    }
}

@Composable
private fun CookStep(
    index: Int,
    text: String,
    amounts: List<StepAmounts.Part>?,
    status: StepStatus,
    timerSeconds: Int?,
    timerWords: LanguageWords,
    timer: StepTimer?,
    isLast: Boolean,
    actions: RecipeActions
) {
    val colors = MaterialTheme.colorScheme
    val body = MaterialTheme.typography.bodyLarge

    if (status == StepStatus.CURRENT) {
        Column(
            Modifier
                .fillMaxWidth()
                .border(BorderStroke(2.dp, colors.primary), RoundedCornerShape(18.dp))
                .padding(20.dp)
        ) {
            Text(
                stringResource(R.string.cook_step_label, index + 1),
                style = MaterialTheme.typography.labelSmall,
                color = colors.tertiary
            )
            Spacer(Modifier.height(8.dp))
            Text(stepText(text, amounts, colors.tertiary), style = body.copy(fontSize = 21.sp, lineHeight = 30.sp))
            if (timerSeconds != null || timer != null) {
                Spacer(Modifier.height(16.dp))
                CurrentTimer(index, timerSeconds, timerWords, timer, actions)
            }
            Spacer(Modifier.height(20.dp))
            Button(
                onClick = actions.onStepDone,
                shape = RoundedCornerShape(14.dp),
                modifier = Modifier.fillMaxWidth().height(52.dp)
            ) {
                Text(
                    stringResource(if (isLast) R.string.cook_done_finish else R.string.cook_done_next),
                    style = MaterialTheme.typography.labelLarge
                )
            }
        }
        return
    }

    val done = status == StepStatus.DONE
    Column(
        Modifier
            .fillMaxWidth()
            .clickable(role = Role.Button, onClickLabel = stringResource(R.string.cd_go_to_step, index + 1)) {
                actions.onStepSelected(index)
            }
            .padding(horizontal = 4.dp, vertical = 6.dp)
            .alpha(if (done) 0.45f else 1f)
    ) {
        Row {
            Text(
                "${index + 1}",
                style = MaterialTheme.typography.titleMedium,
                color = if (done) colors.onSurfaceVariant else colors.tertiary,
                modifier = Modifier.width(32.dp)
            )
            Text(
                // A done step is dimmed as a whole; its amounts keep only their weight.
                stepText(text, amounts, if (done) null else colors.tertiary),
                style = body.copy(
                    textDecoration = if (done) TextDecoration.LineThrough else TextDecoration.None
                ),
                color = colors.onSurfaceVariant
            )
        }
        // A timer still running on a step you've moved on from stays visible: steps overlap.
        if (timer != null) {
            val clockOrDone = if (timer.finished) stringResource(R.string.timers_up)
            else StepTimers.clock(timer.remainingSeconds)
            Text(
                stringResource(R.string.timer_running, clockOrDone),
                style = MaterialTheme.typography.labelLarge.copy(fontFeatureSettings = "tnum"),
                color = colors.tertiary,
                modifier = Modifier.padding(start = 32.dp, top = 4.dp)
            )
        }
    }
}

@Composable
private fun CurrentTimer(
    step: Int,
    timerSeconds: Int?,
    timerWords: LanguageWords,
    timer: StepTimer?,
    actions: RecipeActions
) {
    val colors = MaterialTheme.colorScheme
    if (timer == null) {
        if (timerSeconds == null) return
        OutlinedButton(
            onClick = { actions.onTimerStart(step) },
            shape = RoundedCornerShape(12.dp),
            border = BorderStroke(1.dp, colors.outline)
        ) {
            Text(
                stringResource(R.string.timer_start, StepTimers.label(timerSeconds, timerWords)),
                style = MaterialTheme.typography.labelLarge
            )
        }
        return
    }

    Row(verticalAlignment = Alignment.CenterVertically) {
        Column(Modifier.weight(1f)) {
            Text(
                StepTimers.clock(timer.remainingSeconds),
                style = MaterialTheme.typography.headlineMedium.copy(fontFeatureSettings = "tnum"),
                color = if (timer.finished) colors.tertiary else colors.onSurface
            )
            if (timer.finished) {
                Text(
                    stringResource(R.string.timers_up),
                    style = MaterialTheme.typography.labelMedium,
                    color = colors.tertiary
                )
            }
        }
        if (!timer.finished) {
            TextButton(onClick = { actions.onTimerToggle(step) }) {
                Text(
                    stringResource(if (timer.running) R.string.action_pause else R.string.action_resume),
                    style = MaterialTheme.typography.labelLarge,
                    color = colors.tertiary
                )
            }
        }
        TextButton(onClick = { actions.onTimerReset(step) }) {
            Text(
                stringResource(R.string.action_reset),
                style = MaterialTheme.typography.labelLarge.copy(fontWeight = FontWeight.Medium),
                color = colors.onSurfaceVariant
            )
        }
    }
}

/** Cook mode holds the screen awake: nobody wants it dimming with flour on their hands. */
@Composable
private fun KeepScreenOn() {
    val window = LocalActivity.current?.window
    DisposableEffect(window) {
        window?.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        onDispose { window?.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON) }
    }
}
